File: 2035-count-sub-islands/count-sub-islands.py
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        rows = len(grid1)
        columns = len(grid1[0])
        visited = set()
        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        count = 0

        def isValid(i, j):
            if (i < 0 or j < 0 or i== rows or j == columns):
                return False
            if (i, j) in visited:
                return False
            if (grid2[i][j] != 1):
                return False
            return True

        

        def has_sub_island(i, j):
            is_sub_island = True
            queue = [(i, j)]

            if (grid1[i][j] != 1):
                is_sub_island = False

            while (queue):
                x, y = queue.pop(0)

                for a, b in directions:
                    new_x = a + x
                    new_y = b + y


                    if (isValid(new_x, new_y)):
                        queue.append((new_x, new_y))
                        visited.add((new_x, new_y))
                        if grid1[new_x][new_y] != 1:
                            is_sub_island = False


            return is_sub_island
                    
    

        for i in range(rows):
            for j in range(columns):
                cell = grid2[i][j]
                if (cell):
                    if isValid(i, j) and has_sub_island(i, j):
                        count += 1

        return count

        

File: 2035-count-sub-islands/count-sub-islands.js
var countSubIslands = function(grid1, grid2) {
    let rows = grid1.length
    let columns = grid2[0].length
    let count = 0

    function isValid(i, j) {
        if (i < 0 || j < 0 || i == rows || j == columns) {
            return false
        }
        return true
    }

    function traverse(i, j) {
        let isSub = true
        if (!isValid(i, j) || grid2[i][j] != 1) {
            return true
        }
        if (grid1[i][j] != 1) isSub = false
        grid2[i][j] = 0
        isSub = traverse(i - 1, j) && isSub
        isSub = traverse(i, j + 1) && isSub
        isSub = traverse(i + 1, j) && isSub 
        isSub = traverse(i, j - 1) && isSub
        return isSub
    }

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            let cell = grid2[i][j]
            if (cell == 1) {
                if (traverse(i, j)) count += 1
            }
        }
    }

    return count
};